src/perception/aruco_processing.cpp

```diff
@@ -0,0 +1,136 @@
+#include "aruco_detect.hpp"
+
+#include "filter.hpp"
+
+// TODO: add cache logic to filter out false positives
+
+/**
+ * Detect fiducials from raw image using OpenCV and calculate their screen space centers.
+ * Maintain the immediate buffer - holds all fiducials seen currently on screen.
+ * Later camera space pose information is filled in when we receive point cloud data.
+ *
+ * @param msg
+ */
+void FiducialsNode::imageCallback(sensor_msgs::ImageConstPtr const& msg) {
+    if (!mEnableDetections) return;
+
+    if (mIsVerbose) {
+        ROS_INFO("Got image %d", msg->header.seq);
+    }
+
+    try {
+        mCvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+
+        // Detect the fiducial vertices in screen space and their respective ids
+        cv::aruco::detectMarkers(mCvPtr->image, mDictionary, mCornersCache, mIdsCache, mDetectorParams);
+
+        // Save fiducials currently on screen to the immediate buffer
+        for (size_t i = 0; i < mIdsCache.size(); ++i) {
+            int id = mIdsCache[i];
+            cv::Point2f imageCenter = std::accumulate(mCornersCache[i].begin(), mCornersCache[i].end(), cv::Point2f{}) / 4.0f;
+            mImmediateFiducials[id].id = id;
+            mImmediateFiducials[id].imageCenter = imageCenter;
+        }
+
+        // Remove fiducials in the immediate buffer that are no longer in sight
+        auto it = mImmediateFiducials.begin();
+        while (it != mImmediateFiducials.end()) {
+            if (std::find(mIdsCache.begin(), mIdsCache.end(), it->first) == mIdsCache.end()) {
+                it = mImmediateFiducials.erase(it);
+            } else {
+                ++it;
+            }
+        }
+
+        fiducial_msgs::FiducialTransformArray fidArray{};
+        fidArray.header.frame_id = ODOM_FRAME;
+        fidArray.header.stamp = ros::Time::now();
+        fidArray.header.seq = mSeqNum;
+        fidArray.transforms.reserve(mPersistentFiducials.size());
+
+        // Add readings to the persistent representations of the fiducials
+        for (auto [id, immediateFid]: mImmediateFiducials) {
+            PersistentFiducial& fid = mPersistentFiducials[id];
+            if (!immediateFid.fidInCam.has_value()) continue; // This is set if the point cloud had a valid reading for this fiducial
+
+            fid.id = id;
+            SE3 fidInOdom = SE3::transform(mTfBuffer, ROVER_FRAME, ODOM_FRAME, immediateFid.fidInCam.value(), mSeqNum);
+            fid.setFilterParams(mFilterCount, mFilterProportion);
+            fid.addReading(fidInOdom);
```

Comment:
so because of the fact that the thing that populates the odom reading is async, you have to be careful about this because you might end up adding the same stale reading repeatedly if no reading comes in. Is that okay? It will cause convergence to the last known state as opposed to filtering over the last N known states like you typically intend.

Reply:
Hmm this is a good point I will check the logic about this